File: public-payload/shopping-app/lib/crawl/url-validation.ts
const X_HOSTS = new Set(["x.com", "www.x.com", "twitter.com", "www.twitter.com"]);
const RESERVED_SCREEN_NAMES = new Set([
  "home",
  "explore",
  "search",
  "notifications",
  "messages",
  "settings",
  "compose",
  "intent",
  "share",
  "i",
]);

/** X/TwitterのプロフィールURLだけを正規化して返す。投稿URLやルートURLは拒否する。 */
export function normalizeXProfileUrl(value: unknown): string | null {
  if (typeof value !== "string") return null;
  const raw = value.trim();
  if (!raw) return null;

  let parsed: URL;
  try {
    parsed = new URL(raw.startsWith("http://") || raw.startsWith("https://") ? raw : `https://${raw}`);
  } catch {
    return null;
  }
  if (!X_HOSTS.has(parsed.hostname.toLowerCase())) return null;
  const parts = parsed.pathname.split("/").filter(Boolean);
  if (parts.length !== 1) return null;
  const screenName = parts[0];
  if (!/^[A-Za-z0-9_]{1,15}$/.test(screenName)) return null;
  if (RESERVED_SCREEN_NAMES.has(screenName.toLowerCase())) return null;
  return `https://x.com/${screenName}`;
}

export function extractXScreenName(value: unknown): string | null {
  const normalized = normalizeXProfileUrl(value);
  return normalized ? new URL(normalized).pathname.slice(1) : null;
}

export function normalizeXPostUrl(value: unknown): string | null {
  if (typeof value !== "string") return null;
  try {
    const parsed = new URL(value.trim());
    if (!X_HOSTS.has(parsed.hostname.toLowerCase())) return null;
    const match = parsed.pathname.match(/^\/([A-Za-z0-9_]{1,15})\/status(?:es)?\/(\d+)\/?$/i);
    return match ? `https://x.com/${match[1]}/status/${match[2]}` : null;
  } catch {
    return null;
  }
}

File: src/utils/url_validation.py
"""外部プロフィールURLの検証・正規化。"""

import re
from typing import Optional
from urllib.parse import urlparse


_TWITTER_PROFILE_RE = re.compile(r"^/?@?([A-Za-z0-9_]{1,15})/?$")
_RESERVED_TWITTER_PATHS = {
    "home",
    "explore",
    "i",
    "intent",
    "messages",
    "notifications",
    "search",
    "settings",
    "share",
}


def normalize_twitter_profile_url(value: object) -> Optional[str]:
    """X/TwitterのユーザープロフィールURLだけを正規化して返す。"""
    if not isinstance(value, str):
        return None

    raw = value.strip()
    if not raw:
        return None

    try:
        parsed = urlparse(raw)
    except ValueError:
        return None

    if parsed.scheme.lower() not in {"http", "https"}:
        return None
    if parsed.netloc.lower().split(":", 1)[0] not in {
        "x.com",
        "www.x.com",
        "twitter.com",
        "www.twitter.com",
    }:
        return None
    match = _TWITTER_PROFILE_RE.fullmatch(parsed.path)
    if not match:
        return None

    username = match.group(1)
    if username.lower() in _RESERVED_TWITTER_PATHS:
        return None
    return f"https://x.com/{username}"
